sdk/typescript/src/transactions/ObjectCache.ts

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { bcs } from '../bcs/index.js';
+import type { SuiClient } from '../client/client.js';
+import type { ExecuteTransactionBlockParams } from '../client/index.js';
+import type { Signer } from '../cryptography/keypair.js';
+import { normalizeSuiAddress } from '../utils/sui-types.js';
+import type { OpenMoveTypeSignature } from './blockData/v2.js';
+import type { TransactionBlock } from './TransactionBlock.js';
+import type { TransactionBlockDataResolverPlugin } from './TransactionBlockDataResolver.js';
+
+export interface AsyncCache<T> {
+	get(key: string): Promise<T | null>;
+	set(key: string, value: T): Promise<void>;
+	delete(key: string): Promise<void>;
+	clear(): Promise<void>;
+}
+
+interface ObjectCacheEntry {
+	objectId: string;
+	version: string;
+	digest: string;
+	owner: string | null;
+	initialSharedVersion: string | null;
+}
+
+interface MoveFunctionEntry {
+	package: string;
+	module: string;
+	function: string;
+	parameters: OpenMoveTypeSignature[];
+}
+
+class InMemoryCache<T> implements AsyncCache<T> {
+	#cache = new Map<string, T>();
+
+	async get(key: string): Promise<T | null> {
+		return this.#cache.get(key) ?? null;
+	}
+
+	async set(key: string, value: T): Promise<void> {
+		this.#cache.set(key, value);
+	}
+
+	async delete(key: string): Promise<void> {
+		this.#cache.delete(key);
+	}
+
+	async clear(): Promise<void> {
+		this.#cache.clear();
+	}
+
+	static createCache<T>(): AsyncCache<T> {
+		return new InMemoryCache<T>();
+	}
+}
+
+interface ObjectCacheOptions {
+	createCache?: <T>(name: string) => AsyncCache<T>;
+}
+
+export class ObjectCache implements TransactionBlockDataResolverPlugin {
+	#createCache: <T>(name: string) => AsyncCache<T>;
+	#objects: AsyncCache<ObjectCacheEntry>;
+	#functions: AsyncCache<MoveFunctionEntry>;
+
+	constructor({ createCache = InMemoryCache.createCache }: ObjectCacheOptions = {}) {
+		this.#createCache = createCache;
+		this.#objects = this.#createCache('objects');
```

Comment:
I'm not sure how I feel about this vs just having the cache get this as an argument. Seems easier to do it that way

Reply:
I think combined with the previous comment about knowing what you are caching, having one cache that accepts type, key, value is probably better, then you don't need multiple caches, and can pass in a cache object rather than a function to create a cache